Toggle stream settings

kaustubh-nair opened the pull request:
This PR adds checkboxes to `StreamInfoView` for toggling muted and pinned settings for streams.

There are two parts to this PR and a bunch of minor refactors:
**1. Add support for muting** - We already handle mute events, so all we have to do here is add a checkbox that will toggle mute status.
**2. Add support for pinning** - First support for pinning events are added, then a checkbox is added to `StreamInfoView`.

neiljp commented:
@kaustubh-nair For splitting the second commit, I'd suggest we add the UI change last. Just supporting the subscription events would be a good start - eg. with changes via the webapp.

The muting event change looks pretty good; I'll have another more detailed look later.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -950,14 +950,29 @@ def keypress(self, size: urwid_Size, key: str) -> str:
+        width = max(len(desc) + 2, len(title) + 4, 20)
+        self.controller = controller
+        self.stream_id = stream_id
+        muted_setting = urwid.CheckBox(label="Mute stream",
```
I know this is the label used in the webapp, but it does read a bit strangely to me for some reason. Consistency with the webapp may be important, but this seems like an 'action' rather than a status. Maybe something to think about for a v2, or discuss with frontend.

One idea would be 'muted', which leaves out 'Stream', but perhaps we want a 'Title' above the checkboxes in any case, ie. something like 'Stream settings'?

kaustubh-nair replied:
Yes, a status message would be more appropriate in case of a checkbox.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -950,14 +950,29 @@ def keypress(self, size: urwid_Size, key: str) -> str:
-        super().__init__(controller, stream_info_content, 'STREAM_DESC', width,
-                         title)
+        # 20 is the length required to show stream settings labels.
+        width = max(len(desc) + 2, len(title) + 4, 20)
```
I'm not sure how you plan to develop this further, but it may be useful to simply extract the label text used later and place a `len` expression in here, so the code will express what is happening rather than needing another comment.

kaustubh-nair replied:
Hmm, I still need a comment to explain why I'm adding 4

preetmishra replied:
@kaustubh-nair Which 4?

kaustubh-nair replied:
Ah, sorry this is outdated
This is what I'm referring to
```python3
        # Add 4 to stream settings label for accomodating checkbox.
        width = max(len(desc) + 2, len(title) + 4,
                    len(muted_setting.label) + 4)
```

preetmishra replied:
The +4 is `len([UNICODE]space)`.

Also, there is a typo in the comment.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -950,14 +950,29 @@ def keypress(self, size: urwid_Size, key: str) -> str:
+        muted_setting = urwid.CheckBox(label="Mute stream",
+                                       state=controller.model.is_muted_stream(
+                                           stream_id),
+                                       checked_symbol='✓')
```
I like this, though this may be something we want to standardize on throughout rather than hard-coding. You might like to include that change in a commit with things like extracting the stream 'triangle' and a few other things we've discussed? Maybe in a file in `config`?

kaustubh-nair replied:
Added stream header indicator for now, I'll make a new issue for moving the rest of them

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -950,14 +950,29 @@ def keypress(self, size: urwid_Size, key: str) -> str:
+        urwid.connect_signal(muted_setting, 'change', self.toggle_mute_status)
+
+        # Include blank line between description and stream settings
+        self.stream_info_content = [urwid.Text(desc, align='center'),
```
Do we need this `self`?

kaustubh-nair replied:
No, removed.

kaustubh-nair replied:
I need it while testing

kaustubh-nair replied:
To clarify, I meant while writing the tests.
I need a way to obtain the checkbox object while testing it

Refer to tests above in the same commit
```python3
mute_checkbox = self.stream_info_view.stream_info_content[4]
```

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -950,14 +950,29 @@ def keypress(self, size: urwid_Size, key: str) -> str:
 class StreamInfoView(PopUpView):
     def __init__(self, controller: Any, color: str,
-                 desc: str, title: str) -> None:
+                 desc: str, stream_id: int, title: str) -> None:
```
Possibly for a separate refactor later (or before this), we could consider just passing in the `stream_id`, since we're already calling into the `model` to activate the muting, so could call in to fetch whatever details are required. We don't actually use the color right now.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -950,14 +950,29 @@ def keypress(self, size: urwid_Size, key: str) -> str:
+        super().__init__(controller, self.stream_info_content, 'STREAM_DESC',
+                         width, title)
+
+    def toggle_mute_status(self, button: Any, new_state: bool) -> None:
```
`new_state` is unused; did this code come from somewhere else? Should we name it accordingly or elide it, if it's unused?

kaustubh-nair replied:
It's urwids default function signature for callback functions

neiljp commented on tests/ui/test_ui_tools.py:
```diff
@@ -1338,6 +1340,15 @@ def test_keypress_navigation(self, mocker,
+        toggle_mute_status = self.controller.model.toggle_stream_muted_status
+        stream_id = self.stream_info_view.stream_id
+        size = (10, )
+        mute_checkbox.keypress(size, key)
```
Blank lines around this line, ideally.

kaustubh-nair commented:
@neiljp Thanks for the review! I updated the PR, and added some refactors as you suggested.
I split the pin/unpin commit - though there are some weird bugs that I can't seem to fix. I'll probably have a look tomorrow.

kaustubh-nair commented:
Fixed the bugs, this is ready for review!

kaustubh-nair commented on zulipterminal/helper.py:
```diff
@@ -421,7 +421,8 @@ def classify_unread_counts(model: Any) -> UnreadCounts:
             unread_counts['streams'][stream_id] += count
-        unread_counts['all_msg'] += count
+        if stream_id not in model.muted_streams:
+            unread_counts['all_msg'] += count
```
I didn't use `model.is_muted_stream()` because I can't figure out a way to mock it during testing, without rewriting the whole test.

neiljp replied:
As per another comment, is this a bugfix for muted streams?

kaustubh-nair commented on tests/ui/test_ui_tools.py:
```diff
@@ -2274,9 +2274,8 @@ class TestStreamButton:
-        mocker.patch(STREAMBUTTON + ".mark_muted")
         controller = mocker.Mock()
-        controller.model.muted_streams = {}
+        controller.model.is_muted_stream.return_value = False
```
I'm not sure why this works, but `mocker.patch('controller', 'model.is_muted', return_value=False) doesn't.

preetmishra commented on zulipterminal/config/unicode_characters.py:
```diff
@@ -0,0 +1 @@
+BLACK_RIGHT_POINTING_TRIANGLE = '▶'
```
We may want to use names that are more descriptive so that it is easier for our end-users to depict what modifies which UI element. Also, you could include other Unicode characters that are present in the app.

See also [#zulip-terminal > Unicode characters](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Unicode.20characters/near/862659).

kaustubh-nair replied:
@preetmishra We could do that but I thought it would be better to stick to standard names from wikipedia?
The problem with names related to UI elements is that the same characters may be reused in a lot of places, for e.g. lineboxes.

neiljp replied:
@kaustubh-nair There are quite a few cases not included in commit - I thought you were going to put this into a separate PR/issue?

I agree with @preetmishra that we probably want semantic names rather than direct ones, ie. here the equivalent of 'stream-topic-separator' rather than the currently-used version. That would mean we could change in one place and all the code+tests should update, and ZT run with different characters.

kaustubh-nair replied:
There are a lot of unicode characters in lineboxes so it will be a little time-consuming, hence a new PR would be better.

kaustubh-nair replied:
@neiljp How would you specify semantic names for lines in lineboxes that uses a different style in each corner?
Also the same lines are used as pane seperators, hence I thought a general name would be better.

kaustubh-nair replied:
I intend to remove this commit, opened #717

neiljp commented on tests/helper/test_helper.py:
```diff
@@ -190,13 +190,13 @@ def test_powerset(iterable, map_func, expected_powerset):
     ([99], [['Some general stream', 'Some general unread topic']], {
-        'all_msg': 9,
+        'all_msg': 8,
         'streams': {99: 1},
         'unread_topics': {(99, 'Some private unread topic'): 1},
         'all_mentions': 0,
     }),
     ([1000], [['Secret stream', 'Some private unread topic']], {
-        'all_msg': 11,
+        'all_msg': 8,
```
If these two values vary, is this really a bugfix rather than refactor?

kaustubh-nair replied:
This is testing `classify_unread_counts`, which as per old behavior, did not update `all_msgs` for muted streams. This was done instead in the UI while generating stream buttons. 
This commit moved that part into `classify_unread_counts`, hence there's a change in the values returned.
The overall external behavior did not change, so I'm not sure if it's correct to call it a bugfix

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -164,7 +164,7 @@ def __init__(self, properties: List[Any],
         # Mark muted streams 'M' during button creation.
         if self.model.is_muted_stream(self.stream_id):
-            self.mark_muted()
+            self.update_widget('M')
```
Did the model changes expose something different here?

Looking at the unread counts handling being here, I'd love to move that into a model method to mark a stream/topic muted; it's rather hacky being in UI code.

kaustubh-nair replied:
Yeah, that seems like a useful refactor too.

kaustubh-nair replied:
Unrelated, so opened #722

kaustubh-nair commented on zulipterminal/ui_tools/views.py:
```diff
@@ -949,11 +949,13 @@ def keypress(self, size: urwid_Size, key: str) -> str:
         # Add 4 (for 2 Unicode characters on either side) to the popup title
         # length to make sure that the title gets displayed even when the
         # content is shorter than the title length (+4 Unicode characters).
+        self.controller = controller
```
Used in the callback, moved to next commit.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -951,12 +962,16 @@ def keypress(self, size: urwid_Size, key: str) -> str:
-        width = max(len(desc) + 2, len(title) + 4)
-        stream_info_content = [urwid.Text(desc, align='center')]
-        super().__init__(controller, stream_info_content, 'STREAM_DESC', width,
+        popup_width = max(len(desc) + 2, len(title) + 4)
```
Ah, my bad. I am fixing that in my WIP PR.

kaustubh-nair commented on zulipterminal/model.py:
```diff
@@ -606,6 +606,18 @@ def toggle_stream_muted_status(self, stream_id: int) -> bool:
+    def is_pinned_stream(self, stream_id: int) -> bool:
+        return stream_id in list(stream[1] for stream in self.pinned_streams)
+
+    def toggle_stream_pinned_status(self, stream_id: int) -> bool:
```
When I add all of them, I usually exceed the line limit so I just add the file with the major changes.

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -982,12 +993,16 @@ def keypress(self, size: urwid_Size, key: str) -> str:
-        width = max(len(desc) + 2, len(title) + 4)
-        stream_info_content = [urwid.Text(desc, align='center')]
-        super().__init__(controller, stream_info_content, 'STREAM_DESC', width,
+        popup_width = max(len(desc) + 2, len(title) + 4)
```
@kaustubh-nair Note that the +2 is extra width which we had was for the center alignment. We might want to remove it as we're not using the alignment with `Text`.

preetmishra replied:
(See also https://github.com/zulip/zulip-terminal/pull/695#discussion_r456789654.)

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -982,12 +993,16 @@ def keypress(self, size: urwid_Size, key: str) -> str:
     def __init__(self, controller: Any, color: str,
                  desc: str, title: str) -> None:
+        stream_info_content = [('', [desc])]
+        _, column_widths = self.calculate_table_widths(
```
Could we accept the `popup_width` from here and other widths to it later in the subsequent commits? That would avoid the calculation that you're doing in https://github.com/zulip/zulip-terminal/pull/695/commits/89e53d41bc190524e256aeac06fffed1446a4842#diff-5b93264509b3ec67b9370b2bcbaa7798R1002.

kaustubh-nair replied:
Yeah, that sounds like a good idea, updated :+1:

kaustubh-nair commented on zulipterminal/ui_tools/views.py:
```diff
@@ -782,6 +782,16 @@ def topics_view(self, stream_button: Any) -> Any:
         return w
 
+    def update_structure(self) -> None:
+        if self.is_in_topic_view:
+            return
```
Fixed

neiljp replied:
Could we retain focus on that stream too?

kaustubh-nair commented on zulipterminal/model.py:
```diff
@@ -57,6 +57,13 @@ class ServerConnectionFailure(Exception):
     pass
 
 
+def sort_streams(streams: List[List[str]]) -> None:
+    """
+    Used for sorting model.pinned_streams and model.unpinned_streams.
+    """
+    streams.sort(key=lambda s: s[0].lower())
+
+
```
I hope it is okay to put it here because I can't make it a class method. I don't see a reason to make this a static method, so this was the only choice.
Also, I don't think I understand why `_stream_info_from_subscriptions()` is a static method? Why would you need static methods at all when there is only one `Model` instance?

neiljp replied:
This is fine here for now. I'm expecting there may be more refactoring of stream data in the near future, so just having it local to the model is sufficient.

A static method is just in a different namespace. I think the only difference we could make would be to have it with a leading underscore, but we're not really consistent otherwise right now.

I think `_stream_info_from_subscriptions` might be inlined in future. The advantage is that the individual functions can be tested, and static methods are guaranteed not to modify any data in an object, except potentially that passed in - which can be caught with a type checker, ideally.

kaustubh-nair commented on tests/ui/test_ui_tools.py:
```diff
@@ -1417,6 +1421,28 @@ def test_keypress_navigation(self, mocker,
+        pin_checkbox = self.stream_info_view.widgets[4]
+        toggle_pin_status = self.controller.model.toggle_stream_pinned_status
+        stream_id = self.stream_info_view.stream_id
+        size = (200, 20)
```
Yes, these should be the same - checkbox is a box widget here. Updated.

neiljp commented:
@kaustubh-nair I'm merging this now - thanks for working on this and keeping in with the updates! :tada: This definitely addresses the initial design, and I'd love to have this in for the release - including that first commit which we've debated in chat!

The outstanding points I raised are likely:
* coupling model state to the popup checkbox state (whether via toggles or setting absolute state - see above)
* improving stream focus position retention, which already likely exists as #617 

Other than the subscriber count which I mentioned, another extension might be the per-stream notifications? We added support for that recently, so it'd be great to make that configurable, though that could require examining which settings take precedence.
